core/src/main/java/org/elasticsearch/index/shard/ShardSplittingQuery.java

```diff
@@ -84,44 +94,85 @@ public Scorer scorer(LeafReaderContext context) throws IOException {
-                            new RoutingPartitionedDocIdSetIterator(leafReader, visitor));
+                        TwoPhaseIterator twoPhaseIterator =
+                            parentBitSet == null ? new RoutingPartitionedDocIdSetIterator(leafReader, visitor) :
+                                new NestedRoutingPartitionedDocIdSetIterator(leafReader, visitor, parentBitSet);
+                        return new ConstantScoreScorer(this, score(), twoPhaseIterator);
                     } else {
                         // in the _routing case we first go and find all docs that have a routing value and mark the ones we have to delete
                         findSplitDocs(RoutingFieldMapper.NAME, ref -> {
                             int targetShardId = OperationRouting.generateShardId(indexMetaData, null, ref.utf8ToString());
                             return shardId == targetShardId;
-                        }, leafReader, bitSet::set);
+                        }, leafReader, docId -> {
+                            if (includeDoc.test(docId)) {
+                                bitSet.set(docId);
+                            }
+                        });
+
                         // now if we have a mixed index where some docs have a _routing value and some don't we have to exclude the ones
                         // with a routing value from the next iteration an delete / select based on the ID.
                         if (terms.getDocCount() != leafReader.maxDoc()) {
                             // this is a special case where some of the docs have no routing values this sucks but it's possible today
                             FixedBitSet hasRoutingValue = new FixedBitSet(leafReader.maxDoc());
                             findSplitDocs(RoutingFieldMapper.NAME, ref -> false, leafReader,
-                                    hasRoutingValue::set);
+                                docId -> {
+                                    if (includeDoc.test(docId)) {
+                                        hasRoutingValue.set(docId);
+                                    }
+                                });
+
                             findSplitDocs(IdFieldMapper.NAME, includeInShard, leafReader, docId -> {
-                                if (hasRoutingValue.get(docId) == false) {
+                                if (hasRoutingValue.get(docId) == false && includeDoc.test(docId)) {
                                     bitSet.set(docId);
                                 }
                             });
                         }
                     }
+                    if (parentBitSet != null) {
+                        markChildDocs(parentBitSet, bitSet);
+                    }
                 }
+
                 return new ConstantScoreScorer(this, score(), new BitSetIterator(bitSet, bitSet.length()));
             }
-
-
         };
     }
 
+    private void markChildDocs(BitSet parentDocs, BitSet deletedDocs) {
```

Comment:
Why deletedDocs ? It's the opposite, no ? The bitset of the matching parent+children, allDocs ?

Reply:
naming issue, I will fix. In theory it will hold all docs that need to be deleted by the IndexWriter.